### PR TITLE
Kmp branch

### PR DESCRIPTION
**Pull request by anshulontech**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

### Description

Added `Knuth–Morris–Pratt` algorithm in Cpp.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#376

### CODE REVIEW

File: C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+#include <string>
+
+using namespace std;
+
+void computeLPS(string &, vector<int> &);
+void searchByKMP(string &, string &);
+
+int main()
+{
+    string text = "abacaaabaccabacabaabb";
```
Comment (atarax665) [inline C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp]: Dynamic input required.

File: C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            if (j == 0)
+                i++;
+            else
+                j = lps[j - 1];
+        }
+    }
+}
+
+// Function to Compute LPS.
+// LPS = Longest Prefix which also a suffix.
```
Comment (atarax665) [inline C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp]: Use proper comments to explain your code.

File: C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp (hunk 1 above)
Comment (anshulontech) [reply C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp]: @atarax665 Code Updated With Dynamic Input.

File: C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp (hunk 2 above)
Comment (anshulontech) [reply C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp]: @atarax665 Comments Updated.

File: C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    cin >> n;
+
+    while (n)
+    {
+        cout << endl;
+        cout << "Enter the text string:";
+        cin >> text;
+
+        cout << endl
+             << "Enter the pattern string:";
```
Comment (iamrajiv) [inline C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp]: @anshul7sh format this line like this `cout << endl << "Enter the pattern string:";`

**Review by iamrajiv — CHANGES_REQUESTED** [review]

**Review by iamrajiv — CHANGES_REQUESTED** [review]
@anshul7sh move `searchByKMP` and `computeLPS` above the `main` function. I think that will look better.

File: C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp (hunk 3 above)
Comment (anshulontech) [reply C-Plus-Plus/cp/Knuth–Morris–Pratt.cpp]: @iamrajiv Code Updated as mentioned.

**Review by iamrajiv — APPROVED** [review]

**Comment by anshulontech** [comment]
@iamrajiv, @atarax665  if their is no further changes required than please merge this pull request.

**Comment by iamrajiv** [comment]
@all-contributors please add @anshul7sh for code

**Pull request merged**